### PR TITLE
Combine multiple epoll events copies

### PR DESCRIPTION
**Pull request by DarcySail**
Allocate a larger memory buffer and combine multiple copies into one copy,
to reduce the number of copies from kernel memory to user memory.

Signed-off-by: Hang Su <darcy.sh@antfin.com>

### CODE REVIEW

**Comment by DarcySail** [comment]
@googlebot I signed it!

**Comment by hbhasker** [comment]
This looks good to me. I think the allocation overhead should be minimal unless someone is copying out a large number of events. Even then the scratch buffer only reallocates if the next request is bigger than the previously allocated buffer otherwise it reuses the internal buffer. I don't see any significant downsides to doing this.

File: pkg/sentry/syscalls/linux/sys_epoll.go [hunk 1]
```diff
@@ -111,15 +111,16 @@ func copyOutEvents(t *kernel.Task, addr usermem.Addr, e []epoll.Event) error {
...
+	b := t.CopyScratchBuffer(buffLen)
 	for i := range e {
-		usermem.ByteOrder.PutUint32(b[0:], e[i].Events)
-		usermem.ByteOrder.PutUint32(b[4:], uint32(e[i].Data[0]))
-		usermem.ByteOrder.PutUint32(b[8:], uint32(e[i].Data[1]))
-		if _, err := t.CopyOutBytes(addr, b); err != nil {
-			return err
-		}
-		addr += itemLen
+		usermem.ByteOrder.PutUint32(b[i*12:], e[i].Events)
```
Comment (mrahatm) [inline pkg/sentry/syscalls/linux/sys_epoll.go]: Please use itemLen instead of 12.

Comment (DarcySail) [reply pkg/sentry/syscalls/linux/sys_epoll.go]: ok

**Comment by DarcySail** [comment]
> This looks good to me. I think the allocation overhead should be minimal unless someone is copying out a large number of events. Even then the scratch buffer only reallocates if the next request is bigger than the previously allocated buffer otherwise it reuses the internal buffer. I don't see any significant downsides to doing this.

Due to our profiling, the current code is inefficient when facing heavy Network IO task. 
reproducing with redis-benchmark: `redis-benchmark -q -t ping`

We have reason to believe that it is frequent and trivial switching between kernel mode and user mode leads to this performance problem.

I must admit malloc/gc may complain about the patch, but it's still much faster than original code.

**Pull request merged**